### PR TITLE
Implement `gcd` and `lcm`

### PR DESCRIPTION
**Pull request by niteya-shah**
This PR adds the functions gcd and lcm for cupy.

~Currently lcm is using gcd in the kernel itself. I want to test more to see if this is necessary or calling it externally provides the same performance.~
Testing showed that the performance is the same so the implementation is sufficient

### CODE REVIEW

File: tests/cupy_tests/math_tests/test_rational.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+class TestRational(unittest.TestCase):
+
+    @testing.for_int_dtypes(no_bool=True)
+    @testing.numpy_cupy_array_equal()
+    def test_gcd_basic(self, xp, dtype):
+        a = testing.shaped_random((6, 6), xp, dtype)
+        b = testing.shaped_random((6, 6), xp, dtype)
+        return xp.gcd(a, b)
+
+    @testing.for_float_dtypes()
```
Comment (asi1024) [inline tests/cupy_tests/math_tests/test_rational.py]: Add tests for Boolean and complex dtypes.
```suggestion
    @testing.for_dtypes(['?', 'e', 'f', 'd', 'F', 'D'])
```

File: tests/cupy_tests/math_tests/test_rational.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+
+@testing.gpu
+class TestRational(unittest.TestCase):
+
+    @testing.for_int_dtypes(no_bool=True)
+    @testing.numpy_cupy_array_equal()
+    def test_gcd_basic(self, xp, dtype):
+        a = testing.shaped_random((6, 6), xp, dtype)
+        b = testing.shaped_random((6, 6), xp, dtype)
```
Comment (asi1024) [inline tests/cupy_tests/math_tests/test_rational.py]: Feed different seeds to `a` and `b`.
```py
a = testing.shaped_random((6, 6), xp, dtype, seed=0)
b = testing.shaped_random((6, 6), xp, dtype, seed=1)
```

File: cupy/math/rational.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        return in0;
+    }
+
+template <typename T>
+inline __device__ T lcm(T in0, T in1) {
+        T r = gcd(in0, in1);
+        if(r == 0)
+            return 0;
+        else
+            return (in0 * in1)/r;
```
Comment (asi1024) [inline cupy/math/rational.py]: Avoid integer overflow for large inputs.
```suggestion
            return in0 / r * in1;
```

Comment (niteya-shah) [reply cupy/math/rational.py]: Done

File: tests/cupy_tests/math_tests/test_rational.py (hunk 2 above)
Comment (niteya-shah) [reply tests/cupy_tests/math_tests/test_rational.py]: Done

File: tests/cupy_tests/math_tests/test_rational.py (hunk 1 above)
Comment (niteya-shah) [reply tests/cupy_tests/math_tests/test_rational.py]: Done

**Comment by niteya-shah** [comment]
While implementing this I came across some strange behaviour,
```
simple_ufunc = cupy.core.create_ufunc(
    'simple_ufunc',
    ('b->b', 'h->h', 'i->i', 'l->l'),
    'out0 = in0')

a = cupy.array([True, False, False])
b = cupy.array([False, False, True])
simple_ufunc(b)
```
The above works even though we never specify bool in the accepted types list, instead it permits it by typecasting it to the first type in the list(if we had written, h->h before b->b it typecasts to int16 instead of int8). Is this a bug or expected behaviour?

**Comment by asi1024** [comment]
This is the expected behavior.
`create_ufunc` is undocumented, but for example `cupy.subtract` raises `TypeError` for bool inputs. Please check its implementation.

**Comment by niteya-shah** [comment]
The performance benchmark for gcd and lcm are as follows
gcd

> size = 100000
> numpy
> gcd                 : 13360.778 us   +/-687.684 (min:12873.837 / max:16644.847) us
> cupy
> gcd                 :    11.630 us   +/- 0.938 (min:   11.008 / max:   17.261) us    348.258 us   +/- 2.270 (min:  344.192 / max:  356.928) us
> 

> size = 1000000
> numpy
> gcd                 : 131831.420 us   +/-1681.092 (min:129864.742 / max:137775.571) us
> cupy
> gcd                 :    23.921 us   +/-22.992 (min:   12.101 / max:  148.400) us   3245.719 us   +/-20.154 (min: 3233.440 / max: 3344.928) us

> size = 10000000
> numpy
> gcd                 : 1327643.797 us   +/-16034.114 (min:1319980.853 / max:1446820.613) us
> cupy
> gcd                 :    22.656 us   +/-16.292 (min:   13.221 / max:   73.323) us  32120.730 us   +/-18.294 (min:32102.047 / max:32217.663) us

and for lcm

> size = 100000
> numpy
> lcm                 : 14097.452 us   +/-669.122 (min:13646.137 / max:17272.843) us
> cupy
> lcm                 :    11.926 us   +/- 3.329 (min:   10.634 / max:   32.183) us    366.765 us   +/- 3.625 (min:  362.912 / max:  386.304) us

> size = 1000000
> numpy
> lcm                 : 140687.834 us   +/-4234.880 (min:137613.297 / max:154011.181) us
> cupy
> lcm                 :    17.377 us   +/-13.136 (min:   12.605 / max:   88.989) us   3433.353 us   +/-12.278 (min: 3425.824 / max: 3500.032) us

> size = 10000000
> numpy
> lcm                 : 1395407.041 us   +/-11416.222 (min:1389399.977 / max:1476812.210) us
> cupy
> lcm                 :    21.779 us   +/-10.210 (min:   13.778 / max:   71.256) us  34041.828 us   +/-10.758 (min:34025.921 / max:34093.697) us

This averages at a 40 times speed up for both lcm and gcd compared to numpy

**Comment by niteya-shah** [comment]
@asi1024 can you please review ?

File: tests/cupy_tests/math_tests/test_rational.py (hunk 1 above)
Comment (asi1024) [reply tests/cupy_tests/math_tests/test_rational.py]: Add a test for bool dtype.

File: tests/cupy_tests/math_tests/test_rational.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        b = cupy.random.randint(-10, 10, size=(10, 10)).astype(dtype)
+        with pytest.raises(TypeError):
+            cupy.gcd(a, b)
+
+    @testing.for_int_dtypes(no_bool=True)
+    @testing.numpy_cupy_array_equal()
+    def test_lcm_basic(self, xp, dtype):
+        a = testing.shaped_random((6, 6), xp, dtype, seed=2)
+        b = testing.shaped_random((6, 6), xp, dtype, seed=3)
+        return xp.lcm(a, b)
```
Comment (asi1024) [inline tests/cupy_tests/math_tests/test_rational.py]: `lcm(0, 0)` is not tested in this seed. Why not avoid using `shaped_random` and write a test case explicitly?.

File: tests/cupy_tests/math_tests/test_rational.py (hunk 1 above)
Comment (niteya-shah) [reply tests/cupy_tests/math_tests/test_rational.py]: The test for bool fails as mentioned below, even when I specifically add 'no' casting, this fails.
example
```
import numpy
import cupy

for xp in [cupy, numpy]:
    print(xp.__name__)
    a = xp.array([True, False, False])
    b = xp.array([True, False, True])
    print(xp.logaddexp(a, b, casting='no'))
```

> cupy
> [1.693  0.6934 1.313 ]
> numpy
> Traceback (most recent call last):
>   File "/D/work/Python/trial2.py", line 9, in <module>
>     print(xp.logaddexp(a, b, casting='no'))
> TypeError: ufunc 'logaddexp' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''no''

Can you please tell me how else I should add this test because the proposed way typecasts it to int8 and the doesnt give the desired result.

File: tests/cupy_tests/math_tests/test_rational.py (hunk 4 above)
Comment (niteya-shah) [reply tests/cupy_tests/math_tests/test_rational.py]: I have added additional test cases that cover boundary cases -> (0,0), negative and positive, and negative and negative

File: tests/cupy_tests/math_tests/test_rational.py (hunk 1 above)
Comment (asi1024) [reply tests/cupy_tests/math_tests/test_rational.py]: Please refer this line: https://github.com/cupy/cupy/blob/master/cupy/core/_routines_math.pyx#L744

File: tests/cupy_tests/math_tests/test_rational.py (hunk 4 above)
Comment (asi1024) [reply tests/cupy_tests/math_tests/test_rational.py]: What is the role of `test_lcm_basic`? I think it is better to write all the test cases explicitly.

File: tests/cupy_tests/math_tests/test_rational.py (hunk 1 above)
Comment (niteya-shah) [reply tests/cupy_tests/math_tests/test_rational.py]: Thanks a lot, I have modified the ufunc and added the test case.

File: tests/cupy_tests/math_tests/test_rational.py (hunk 4 above)
Comment (niteya-shah) [reply tests/cupy_tests/math_tests/test_rational.py]: I have added the test cases with all values explicitly written and have tried to choose values that represent all the boundary cases.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM. Thanks for your contribution!
